### PR TITLE
update mop-utils

### PR DESCRIPTION
**Pull request by suxi-ms**

### CODE REVIEW

File: packages/mop_utils/mop_utils/base_model_wrapper.py [hunk 1]
```diff
@@ -3,7 +3,8 @@
 
 
 class InferenceInput:
```
[the remark this answers is not in the document]
Comment (suxi-ms) [reply packages/mop_utils/mop_utils/base_model_wrapper.py:5]: done

File: packages/mop_utils/mop_utils/base_model_wrapper.py [hunk 2]
```diff
@@ -104,9 +105,17 @@ def init(self, model_root: str, **kwargs):
...
     @abstractmethod
-    def inference(self, item: InferenceInput, **kwargs) -> InferenceOutput:
+    def inference(self, item: str, **kwargs) -> str:
         raise NotImplementedError("inference() method is not implemented")
 
     @abstractmethod
-    def inference_batch(self, items: List[InferenceInput], **kwargs) -> List[InferenceOutput]:
-        raise NotImplementedError("inference_batch() method is not implemented")
+    def inference_batch(self, item: List[str], **kwargs) -> List[str]:
+        raise NotImplementedError("inference() method is not implemented")
```
[the remark this answers is not in the document]
Comment (suxi-ms) [reply packages/mop_utils/mop_utils/base_model_wrapper.py:113]: done

File: packages/mop_utils/mop_utils/base_model_wrapper.py [hunk 3]
```diff
@@ -104,9 +105,17 @@ def init(self, model_root: str, **kwargs):
         raise NotImplementedError("init() method is not implemented")
 
     @abstractmethod
-    def inference(self, item: InferenceInput, **kwargs) -> InferenceOutput:
+    def inference(self, item: str, **kwargs) -> str:
```
[the remark this answers is not in the document]
Comment (suxi-ms) [reply packages/mop_utils/mop_utils/base_model_wrapper.py:108]: done

File: packages/mop_utils/mop_utils/base_model_wrapper.py [hunk 4]
```diff
@@ -58,7 +59,11 @@ def height(self, height: int) -> None:
...
 
 
-class InferenceOutput:
+class InferenceInput(MopInferenceInput):
+    '''this is a subclass to be compatible with old version'''
+
+
+class MopInferenceOutput:
     def __init__(self, confidence_scores: Optional[Dict] = None, predicted_labels: Optional[Dict] = None) -> None:
         self.__confidence_scores__: Dict = confidence_scores
```
Comment (conhua) [inline packages/mop_utils/mop_utils/base_model_wrapper.py:68]: __probability_scores__

File: packages/mop_utils/mop_utils/base_model_wrapper.py [hunk 5]
```diff
@@ -104,9 +113,17 @@ def init(self, model_root: str, **kwargs):
         raise NotImplementedError("init() method is not implemented")
 
     @abstractmethod
-    def inference(self, item: InferenceInput, **kwargs) -> InferenceOutput:
+    def inference(self, item: dict, **kwargs) -> dict:
```
Comment (conhua) [inline packages/mop_utils/mop_utils/base_model_wrapper.py:116]: Dict

File: packages/mop_utils/mop_utils/inference_wrapper.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    def __init__(self, base_model_wrapper: BaseModelWrapper) -> None:
+        self.model_wrapper = base_model_wrapper
+
+    def init(self, model_root: str) -> None:
+        self.model_wrapper.init(model_root)
+
+    def run(self, item: Dict, triggered_by_mop: False) -> Dict:
+        if not triggered_by_mop:
+            customized_output = self.model_wrapper.inference(item)
+            return [customized_output]
```
Comment (conhua) [inline packages/mop_utils/mop_utils/inference_wrapper.py:34]: Dict?

File: packages/mop_utils/mop_utils/inference_wrapper.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+        else:
+            if isinstance(item, dict):
```
Comment (conhua) [inline packages/mop_utils/mop_utils/inference_wrapper.py:36]: must be Dict?

File: packages/mop_utils/mop_utils/inference_wrapper.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        if isinstance(obj, np.floating):
+            return float(obj)
+        if isinstance(obj, np.ndarray):
+            return obj.tolist()
+        if isinstance(obj, np.bool_):
+            return bool(obj)
+        return json.JSONEncoder.default(self, obj)
+
+
+def mop_init():
```
Comment (conhua) [inline packages/mop_utils/mop_utils/inference_wrapper.py:113]: Add a parameter model_root

File: packages/mop_utils/mop_utils/inference_wrapper.py [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    global batch_model
+    global batch_size
+    inference_wrapper.init(os.path.join(os.getenv("AZUREML_MODEL_DIR"), "../model"))
```
Comment (conhua) [inline packages/mop_utils/mop_utils/inference_wrapper.py:116]: model_root

File: packages/mop_utils/mop_utils/inference_wrapper.py [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+
+    # assign batch_model is enabled
+    dynamic_batch_args = _get_dynamic_batch_args()
+    if dynamic_batch_args is not None:
+        batch_model = DynamicBatchModel(WrapModel(), **dynamic_batch_args)
+        batch_size = dynamic_batch_args.get('max_batch_size')
+
+
+def mop_run(raw_data: any, triggered_by_mop) -> any:
```
Comment (conhua) [inline packages/mop_utils/mop_utils/inference_wrapper.py:125]: Union [List, Dict]

File: packages/mop_utils/mop_utils/inference_wrapper.py [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+
+
+class MOPInferenceWrapper:
+    def __init__(self, base_model_wrapper: BaseModelWrapper) -> None:
+        self.model_wrapper = base_model_wrapper
+
+    def init(self, model_root: str) -> None:
+        self.model_wrapper.init(model_root)
+
+    def run(self, item: Dict, triggered_by_mop: False) -> Dict:
```
Comment (conhua) [inline packages/mop_utils/mop_utils/inference_wrapper.py:31]: bool=False

File: packages/mop_utils/mop_utils/inference_wrapper.py [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+    if isinstance(raw_data, dict):
+        inference_result = inference_wrapper.run(raw_data, triggered_by_mop)
+        return inference_result
+    if isinstance(raw_data, list):
+        inference_result = inference_wrapper.run_batch(raw_data, triggered_by_mop)
+        return inference_result
+    raise Exception("Invalid input data format")
+
+
+def build_response(inference_result: any) -> any:
```
Comment (conhua) [inline packages/mop_utils/mop_utils/inference_wrapper.py:138]: should move to score?

File: packages/mop_utils/mop_utils/inference_wrapper.py [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+    # get args
+    args = {
+        'max_batch_size': dynamic_batch.get('maxBatchSize'),
+        'idle_batch_size': dynamic_batch.get('idleBatchSize'),
+        'max_batch_interval': dynamic_batch.get('maxBatchInterval'),
+    }
+    return {k: v for k, v in args.items() if v is not None}
+
+
+class NumpyJsonEncoder(json.JSONEncoder):
```
Comment (conhua) [inline packages/mop_utils/mop_utils/inference_wrapper.py:100]: move to score?

File: packages/mop_utils/mop_utils/base_model_wrapper.py [hunk 14]
```diff
@@ -104,9 +113,17 @@ def init(self, model_root: str, **kwargs):
...
         raise NotImplementedError("inference() method is not implemented")
 
     @abstractmethod
-    def inference_batch(self, items: List[InferenceInput], **kwargs) -> List[InferenceOutput]:
+    def inference_batch(self, item: List[dict], **kwargs) -> List[dict]:
         raise NotImplementedError("inference_batch() method is not implemented")
+
+    @abstractmethod
+    def convert_mop_input_to_customized_input(self, mop_input: MopInferenceInput, **kwargs) -> dict:
```
Comment (conhua) [inline packages/mop_utils/mop_utils/base_model_wrapper.py:124]: convert_mop_input_to_model_input

File: packages/mop_utils/mop_utils/base_model_wrapper.py [hunk 15]
```diff
@@ -104,9 +113,17 @@ def init(self, model_root: str, **kwargs):
...
+        raise NotImplementedError("convert_mop_input_to_customized_input() method is not implemented")
+
+    @abstractmethod
+    def convert_customized_output_to_mop_output(self, customized_output: dict, **kwargs) -> MopInferenceOutput:
```
Comment (conhua) [inline packages/mop_utils/mop_utils/base_model_wrapper.py:128]: convert_model_output_to_mop_output

File: packages/mop_utils/setup.py [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+
+setup(
+   name='mop_utils',
+   version='1.6',
+   description='Utilities for MOP',
+   packages=['mop_utils'],
+   package_dir={'mop_utils': './mop_utils'},
+   python_requires='>=3.8.0',
+   install_requires = [
+      'azureml-contrib-services==1.48',
```
Comment (conhua) [inline packages/mop_utils/setup.py:12]: no azureml?

File: packages/mop_utils/mop_utils/base_model_wrapper.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (suxi-ms) [reply packages/mop_utils/mop_utils/base_model_wrapper.py]: done

File: packages/mop_utils/mop_utils/base_model_wrapper.py (hunk 14 above)
Comment (suxi-ms) [reply packages/mop_utils/mop_utils/base_model_wrapper.py:124]: done

File: packages/mop_utils/mop_utils/base_model_wrapper.py (hunk 15 above)
Comment (suxi-ms) [reply packages/mop_utils/mop_utils/base_model_wrapper.py:128]: done

File: packages/mop_utils/setup.py (hunk 16 above)
Comment (suxi-ms) [reply packages/mop_utils/setup.py:12]: removed

File: packages/mop_utils/mop_utils/base_model_wrapper.py [hunk 17]
```diff
@@ -104,9 +105,17 @@ def init(self, model_root: str, **kwargs):
...
-        raise NotImplementedError("inference_batch() method is not implemented")
+    def inference_batch(self, item: List[str], **kwargs) -> List[str]:
+        raise NotImplementedError("inference() method is not implemented")
+
+    @abstractmethod
+    def convert_mop_input_to_customized_input(self, mop_input: InferenceInput, **kwargs) -> str:
+        raise NotImplementedError("convert_mop_input_to_customized_input() method is not implemented")
+
+    @abstractmethod
+    def convert_customized_output_to_mop_output(self, customized_output: str, **kwargs) -> InferenceOutput:
```
[the remark this answers is not in the document]
Comment (suxi-ms) [reply packages/mop_utils/mop_utils/base_model_wrapper.py]: done

File: packages/mop_utils/mop_utils/base_model_wrapper.py (hunk 4 above)
Comment (suxi-ms) [reply packages/mop_utils/mop_utils/base_model_wrapper.py:68]: done

**Review by conhua — APPROVED** [review]

**Pull request merged**